File: k_means/go/kmeans.go
package main

import (
	"fmt"
	"math"
	"math/rand"
)

type Point struct {
	X, Y float64
}

type Cluster struct {
	Centroid   Point
	NumPoints  int
}

func distance(p1, p2 Point) float64 {
	return math.Sqrt(math.Pow(p2.X-p1.X, 2) + math.Pow(p2.Y-p1.Y, 2))
}

func initializeCentroids(dataPoints []Point, k int) []Cluster {
	clusters := make([]Cluster, k)

	for i := 0; i < k; i++ {
		randomIndex := rand.Intn(len(dataPoints))
		clusters[i].Centroid = dataPoints[randomIndex]
		clusters[i].NumPoints = 0
	}

	return clusters
}

func assignPointsToCentroids(dataPoints []Point, clusters []Cluster) {
	for i := range dataPoints {
		minDistance := distance(dataPoints[i], clusters[0].Centroid)
		clusterIndex := 0

		for j := 1; j < len(clusters); j++ {
			currDistance := distance(dataPoints[i], clusters[j].Centroid)
			if currDistance < minDistance {
				minDistance = currDistance
				clusterIndex = j
			}
		}

		clusters[clusterIndex].NumPoints++
	}
}

func updateCentroids(dataPoints []Point, clusters []Cluster) {
	for i := range clusters {
		sumX, sumY := 0.0, 0.0
		numPoints := clusters[i].NumPoints

		for j := range dataPoints {
			sumX += dataPoints[j].X
			sumY += dataPoints[j].Y
		}

		clusters[i].Centroid.X = sumX / float64(numPoints)
		clusters[i].Centroid.Y = sumY / float64(numPoints)
	}
}

func kMeans(dataPoints []Point, k int) []Cluster {
	clusters := initializeCentroids(dataPoints, k)
	maxIterations := 100
	iteration := 0

	for iteration < maxIterations {
		assignPointsToCentroids(dataPoints, clusters)
		updateCentroids(dataPoints, clusters)
		iteration++
	}

	return clusters
}

func main() {
	dataPoints := []Point{
		{2.0, 3.0},
		{2.5, 4.5},
		{1.5, 2.5},
		{6.0, 5.0},
		{7.0, 7.0},
		{5.0, 5.5},
		{9.0, 2.0},
		{10.0, 3.5},
		{9.5, 2.5},
	}

	k := 3

	clusters := kMeans(dataPoints, k)

	for i, cluster := range clusters {
		fmt.Printf("Cluster %d: Centroid (%.2f, %.2f), Points: %d\n", i+1, cluster.Centroid.X, cluster.Centroid.Y, cluster.NumPoints)
	}
}

File: k_means/julia/kmeans.jl
struct Point
    x::Float64
    y::Float64
end

struct Cluster
    centroid::Point
    numPoints::Int64
end

function distance(p1::Point, p2::Point)
    return sqrt((p2.x - p1.x)^2 + (p2.y - p1.y)^2)
end

function initializeCentroids(dataPoints, k)
    centroids = Point[]
    for _ in 1:k
        randomIndex = rand(1:length(dataPoints))
        push!(centroids, dataPoints[randomIndex])
    end
    return centroids
end

function assignPointsToCentroids(dataPoints, clusters)
    for point in dataPoints
        minDistance = distance(point, clusters[1].centroid)
        clusterIndex = 1

        for j in 2:length(clusters)
            currDistance = distance(point, clusters[j].centroid)
            if currDistance < minDistance
                minDistance = currDistance
                clusterIndex = j
            end
        end

        clusters[clusterIndex].numPoints += 1
    end
end

function updateCentroids(dataPoints, clusters)
    for cluster in clusters
        sumX = 0.0
        sumY = 0.0
        numPoints = cluster.numPoints

        for point in dataPoints
            sumX += point.x
            sumY += point.y
        end

        cluster.centroid.x = sumX / numPoints
        cluster.centroid.y = sumY / numPoints
    end
end

function kMeans(dataPoints, k)
    clusters = Cluster[]
    for centroid in initializeCentroids(dataPoints, k)
        push!(clusters, Cluster(centroid, 0))
    end

    iteration = 0
    while iteration < 100
        assignPointsToCentroids(dataPoints, clusters)
        updateCentroids(dataPoints, clusters)
        iteration += 1
    end

    return clusters
end

dataPoints = [Point(2.0, 3.0),
              Point(2.5, 4.5),
              Point(1.5, 2.5),
              Point(6.0, 5.0),
              Point(7.0, 7.0),
              Point(5.0, 5.5),
              Point(9.0, 2.0),
              Point(10.0, 3.5),
              Point(9.5, 2.5)]

clusters = kMeans(dataPoints, 3)

for (i, cluster) in enumerate(clusters)
    println("Cluster $i: Centroid (${cluster.centroid.x}, ${cluster.centroid.y}), Points: ${cluster.numPoints}")
end